server/decorators.go

```diff
@@ -38,3 +46,59 @@ func Delay(delay time.Duration, h http.HandlerFunc) http.HandlerFunc {
 		time.Sleep(delay)
 	}
 }
+
+// Cached decorator inspect a Greet resource into a cache to suggests
+// some equivalents when some data are missing
+func Cached(h http.HandlerFunc) http.HandlerFunc {
+	// Create the cache struct
+	c, err := greet.NewCache()
+	if err != nil {
+		panic(err)
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		// decode body request using a temporal reader
+		// and rewriting it to avoid handler chain errors
+		var t greet.Greet
+		b := bytes.NewBuffer(make([]byte, 0))
+		reader := io.TeeReader(r.Body, b)
+		r.Body = ioutil.NopCloser(b)
+		err := json.NewDecoder(reader).Decode(&t)
+		if err != nil {
+			log.Print("Error decoding body: ", err)
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// do handler chain
+		h.ServeHTTP(w, r)
+
+		// inspect cache and append custom messages depending on it status
+		store := c.GetAll()
+		gr, prs := c.SetIfAbsent(t)
+		if prs {
+			log.Printf("Resource %+v cached as: %+v", t, gr)
+			fmt.Fprintf(w, "Hey %s!! Your cached location was: %s\n", t.Name, gr.Location)
+		} else {
+			log.Printf("Resource %+v is not cached", t)
+			// use and slice to resolve suggestions
+			var suggestions = make([]greet.Greet, 0)
+			for _, v := range store {
+				if v.Name != t.Name && strings.Contains(v.Name, t.Name) {
+					suggestions = append(suggestions, v)
+				}
+			}
+			n := len(suggestions)
+			log.Printf("%d suggested resources: %+v", n, suggestions)
+			if n > 0 {
+				if n == 1 {
+					gr = suggestions[0]
+					fmt.Fprintf(w, "Are you %s from %s?\n", gr.Name, gr.Location)
+				} else {
+					fmt.Fprintf(w, "Are you some of these suggestions? %s\n", strings.Trim(fmt.Sprintf("%+v", suggestions), "[]"))
+				}
+			}
```

Comment:
Pregunta!!!

Al estar agregando bytes tras procesarse `h.ServeHTTP(w, r)` no habría que sobreescribir la cabecera `Content-Length` del response resultante?

Como soy nuevo en Go, no me queda nada claro tras leer el [siguiente enlace](https://justinas.org/writing-http-middleware-in-go)

Reply:
Sí, es un issue muy corriente al utilizar esta técnica, y muchos de tus compañeros han tenido que ir sobrescribiendo el response cada vez. Yo lo planteo como un problema de composición: no debería hacer falta el h.ServeHTTP, ya que sería un único handler, que lee y escribe de la caché del middleware que lo envuelve.

Reply:
Aquellos casos que describe el artículo los evito: si tengo que escribir en el response compongo los decoradores de tal forma que no se "pisen" entre sí. La idea de esta técnica es desacoplar acciones unas de otras pero siempre que esto resulte intuitivo y no haya que hacer muchos trucos. Si no, carece de sentido.

Reply:
La técnica de usar un map que identifique cada request para saber qué dato le corresponde es de lo más tricky. Muy chula.

Reply:
Con qué estás comprobando el Content-Length?

Reply:
> Con qué estás comprobando el Content-Length?

De momento con el `resp.ContentLength` en el `./client`, que parece que casa correctamente correctamente con el tamaño de los bytes del body. Debe estar haciendo magia por dentro, al no establecerle las cabeceras desde el server, el client lo autocalcula (que sería lo lógico en cualquier comunicacion HTTP)

Ver: https://github.com/pabloos/GoCourse/pull/16/commits/55d03c8f94066f0f785da1990053b94edce7b111